_posts/2020-02-24-qute.adoc

```diff
@@ -0,0 +1,182 @@
+On the other hand, a user is not protected from tedious errors caused by typos.
+Qute templates can be _optionally_ type-safe.
+What does it actually mean?
+A template may contain one or more *parameter declarations*.
+A parameter declaration binds a concrete type information to a given identifier in the current context.
+And what are the advantages of having a type-safe template?
+
+* Quarkus attempts to _validate all expressions_ that reference a parameter declaration. If an invalid/incorrect expression is found the build fails. 
+
+TIP: In the development mode, all files located in the `src/main/resources/templates` directory are watched for changes and modifications are immediately visible. That also implies that your application fails fast whenever a type error occurs.
+
+* A _value resolver_ is generated for all types used in parameter declarations so that it’s possible to access its properties without reflection. This is very useful when targeting GraalVM native images. 
+* We have few more ideas in our TODO list, such as performance optimizations for type-safe expressions, etc. 
+
+.Type-safe Template Example
+[source,html]
+----
+{@org.acme.Foo foo} <1>
+<!DOCTYPE html>
+<html>
+<head>
+<meta charset="UTF-8">
+<title>Qute Hello</title>
+</head>
+<body>
+  <h1>{title}</h1>  <2>
+  <p>{foo.message}</p>  <3>
+  {#for foo in baz.foos}
+    <p>Hello {foo.message}!</p>  <4>
+  {/for}
+</body>
+</html>
+----
+<1> Parameter declaration - maps `foo` to `org.acme.Foo`.
+<2> `{title}` is not validated - not matching a param declaration.
+<3> `{foo.message}` is validated. `org.acme.Foo` must have a property `message` or a matching template extension method must exist. 
+<4> `{foo.message}` is not validated because `foo` is overridden in the loop section and there is no type information available.
+
+=== First-class Quarkus Citizen
+
+In Quarkus, all templates located in the `src/main/resources/templates` directory are validated and can be easily injected.
+
+.Template Injection Example
+[source,java]
+----
+package org.acme.qute;
+
+import io.quarkus.qute.Template;
+
+class MyBean {
+
+    @Inject 
+    Template items; <1>
+    
+    @Inject
+    Service service;
+
+    String renderItems() {
+       return items.data("items", service.getItems()).render(); <2>
+    }
+}
+----
+<1> The field name is used to locate the template. In this particular case, the container will attempt to locate a template with path `src/main/resources/templates/items.html`. If there is no such template available the build fails.
+<2> See the https://quarkus.io/guides/qute-reference#hello-world-example[Hello World Example] to explore the basic workflow.
+
+Moreover, a preconfigured `Engine` instance is provided and available for injection.
+The `Engine` is a central point for template management and provides some low-level API.
+
+=== RESTEasy Integration
+
+If used together with RESTEasy a special `ContainerResponseFilter` is registered automatically.
```

Comment:
I don’t think you should mention ContainerTesponseFilter. Do people need to know the impl detail ?

Reply:
Good point. I've just copied the text from the ref guide.

Reply:
Hm, actually this is an interface from JAX-RS, not an impl class... but I will remove it anyway.